src/legacy/core_plugins/kibana/public/discover/np_ready/components/table/table.tsx

```diff
@@ -68,11 +69,55 @@ export function DocViewTable({
             const displayNoMappingWarning =
               !mapping(field) && !displayUnderscoreWarning && !isArrayOfObjects;
 
+            // Discover doesn't flatten arrays of objects, so for documents with an `object` or `nested` field that
+            // contains an array, Discover will only detect the top level root field. We want to detect when those
+            // root fields are `nested` so that we can display the proper icon and label. However, those root
+            // `nested` fields are not a part of the index pattern. Their children are though, and contain nested path
+            // info. So to detect nested fields we look through the index pattern for nested children
+            // whose path begins with the current field. There are edge cases where
+            // this could incorrectly identify a plain `object` field as `nested`. Say we had the following document
+            // where `foo` is a plain object field and `bar` is a nested field.
+            // {
+            //   "foo": [
+            //   {
+            //     "bar": [
+            //       {
+            //         "baz": "qux"
+            //       }
+            //     ]
+            //   },
+            //   {
+            //     "bar": [
+            //       {
+            //         "baz": "qux"
+            //       }
+            //     ]
+            //   }
+            // ]
+            // }
+            //
+            // The following code will search for `foo`, find it at the beginning of the path to the nested child field
+            // `foo.bar.baz` and incorrectly mark `foo` as nested. Any time we're searching for the name of a plain object
+            // field that happens to match a segment of a nested path, we'll get a false positive.
+            // We're aware of this issue and we'll have to live with
+            // it in the short term. The long term fix will be to add info about the `nested` and `object` root fields
+            // to the index pattern, but that has its own complications which you can read more about in the following
+            // issue: https://github.com/elastic/kibana/issues/54957
+            const isNestedField =
+              !indexPattern.fields.find(patternField => patternField.name === field) &&
+              !!indexPattern.fields.find(patternField => {
+                // We only want to match a full path segment
+                const nestedRootRegex = new RegExp(escapeRegExp(field) + '(\\.|$)');
+                return nestedRootRegex.test(patternField.subType?.nested?.path ?? '');
+              });
+            const fieldType = isNestedField ? 'nested' : undefined;
```

Comment:
I think we could use `field.type` here instead of `undefined`. Since the `FieldName` component later looks first on `field.type` and then on the `fieldType` property (which we could discuss that this might not be ideal), I think we want to pass the correct `fieldType` no matter if nested or not here, instead just passing it if it's `nested`. It would still work, since in case we pass `nested` we know that `field.type` is `undefined`, but it kind of sounds a bit like using a side effect for me.... so long story short: could we just make sure `fieldType` here is always set correctly :D

Reply:
Works for me, I went back and forth on this myself during implementation.

https://github.com/elastic/kibana/pull/54978/commits/a6a5aa89a5e14cab6c7bea840ba74273d9a4586e